Flavor migration and unit tests

meoflynn opened the pull request:
### Description:

Python Script in the Lib layer for OpenStack Flavors for migrating flavors between Clouds. 
<!--
This should be a brief one or two line description of the PR. Details should be contained in commit messages.
-->

### Special Notes:

<!-- This section and header can be removed if not required.

Examples of special notes that must be included in the PR:
- Changes any parameter names, or allowed values
- Renames any actions, workflows or sensors
- Requires any additional config files placed onto the server

Or anything else you may want to note:

-->

---

### Submitter:

Have you (where applicable):

* [x] Added unit tests?
* [ ] Checked the latest commit runs on Dev?
* [ ] Updated the example config file(s) and README?

---

### Reviewer

Does this PR:

* [ ] Place non-StackStorm code into the `lib` directory?
* [ ] Have unit tests for the action/sensor and `lib` layers?
* [ ] Have clear and obvious action parameter names and descriptions?

meoflynn committed:
Flavor migration and unit tests

Python script and unit test in the lib layer for migrating OpenStack Flavors

DavidFair commented on lib/openstack_api/openstack_flavor.py:
```diff
@@ -0,0 +1,127 @@
+        elif isinstance(flavor_id, Flavor):
+            extra_specs = flavor_id.extra_specs
+        else:
+            extra_specs = {}
```
Could we change this to:

```
        if isinstance(flavor_id, str):
            flavor = self.get_flavor(cloud_account, flavor_id)
            extra_specs = flavor.extra_specs
        else:
            extra_specs = flavor_id.extra_specs
```

Since we should only get a flavor or string type here, so the `else {}` is a subtle bug waiting to happen

DavidFair commented on lib/openstack_api/openstack_flavor.py:
```diff
@@ -0,0 +1,127 @@
+        :param extra_specs: Dictionary of extra specs for the flavor
+        """
+        with self._connection_cls(cloud_account) as conn:
+            if extra_specs != {}:
```
We can rely on Python truthy types:
```
if extra_specs:
   conn.set_flavor_specs(...)
```

DavidFair commented on lib/openstack_api/openstack_flavor.py:
```diff
@@ -0,0 +1,127 @@
+        """
+        missing_flavors = self.get_missing_flavors(source_cloud, dest_cloud)
+        # if there are missing flavors in the list i.e. missing_flavors is not an empty list
+        if missing_flavors:
```
Can we change this so the main branch isn't nested in the if-condition:

```python
if not missing_flavors:
    return

for flavor_name in missing_flavors:
    ....
```

DavidFair commented on tests/lib/test_openstack_flavor.py:
```diff
@@ -0,0 +1,140 @@
+        self.api.get_flavor.assert_called_once_with(mocked_flavor_name)
+        self.mocked_connection.assert_called_once_with(mocked_cloud_name)
+
+    def test_get_missing_flavors_calls_list_flavor(self):
```
`test_get_missing_flavors_calls_list_flavor_with_correct_clouds`
And the docstring is incorrect, as this test is checking we pass the source and dest cloud in the correct order rather than checking the diff in the next test

DavidFair commented on tests/lib/test_openstack_flavor.py:
```diff
@@ -0,0 +1,140 @@
+        """
+        Test getting the extra specs from a flavor
+        """
+        mocked_cloud_name = NonCallableMock()
+        mocked_flavor = NonCallableMock()
+
+        self.instance.get_flavor(mocked_cloud_name, mocked_flavor)
+        self.instance.get_flavor_specs(mocked_cloud_name, mocked_flavor)
```
This test has no asserts, so it's not testing anything at the moment?

DavidFair commented on tests/lib/test_openstack_flavor.py:
```diff
@@ -0,0 +1,140 @@
+        flavor_id = NonCallableMock()
+        extra_specs = NonCallableMock()
+
+        self.instance.set_flavor_specs(mocked_cloud_name, flavor_id, extra_specs)
+        self.mocked_connection.assert_called_once_with(mocked_cloud_name)
```
Ditto, we're checking the connection opens but not the set_specs method is called

DavidFair commented on tests/lib/test_openstack_flavor.py:
```diff
@@ -0,0 +1,140 @@
+        source_flavor = Mock()
+        dest_flavor = self.instance.create_flavor(mocked_dest_cloud, source_flavor)
+        assert dest_flavor == self.api.create_flavor.return_value
+        self.instance.get_flavor_specs(mocked_source_cloud, dest_flavor.id)
+        self.instance.set_flavor_specs(
```
Are these missing `.assert_called_once_with`? as it's making the method calls manually 

Also it doesn't look like `migrate_flavor` is actually called in this test?

meoflynn committed:
Add return None to migrate_flavors method when there are no missing flavors

DavidFair commented on lib/openstack_api/openstack_flavor.py:
```diff
@@ -0,0 +1,132 @@
+        :param dest_cloud: Cloud account for destination cloud
+        :return: List of names of missing flavors or empty list
+        """
+        # list flavors from prod
```
Nit: this should say from source cloud

DavidFair commented on lib/openstack_api/openstack_flavor.py:
```diff
@@ -0,0 +1,132 @@
+        # list flavors from prod
+        source_flavors = self.list_flavor(source_cloud)
+
+        # list flavors from dev
```
^

DavidFair commented on lib/openstack_api/openstack_flavor.py:
```diff
@@ -0,0 +1,132 @@
+        dest_flavors = self.list_flavor(dest_cloud)
+
+        # Prepare list of flavors that are in prod and dev
+        source_flavor_names = []
```
We can remove the two for loops below and use list comprehension instead 
```
source_flavor_names = [source.name for source in source_flavors]
dest_flavor_names = [dest.name for dest in dest_flavors]
```

Also we can go straight to a set rather than doing list -> set
```
source_flavor_names = {source.name for source in source_flavors}
dest_flavor_names = {dest.name for dest in dest_flavors}
```

This allows us to get rid of the `set(source_flavor_names)` and `set(dest_flavor_names)` casts below

DavidFair commented on lib/openstack_api/openstack_flavor.py:
```diff
@@ -0,0 +1,132 @@
+        """
+
+        if (
+            flavor_data.swap == ""
```
Rely on the falsey type check for this:
```
if (flavor_data.swap):
   flavor_swap = flavor_data.swap
else:
   flavor_swap = 0
```

DavidFair commented on tests/lib/test_openstack_flavor.py:
```diff
@@ -0,0 +1,211 @@
+        mocked_name = NonCallableMock()
+        flavors = self.instance.list_flavor(mocked_name)
+        assert flavors == self.api.list_flavors.return_value
+        self.mocked_connection.assert_called_once_with(mocked_name)
```
Nit we don't check `self.api.list_flavors.assert_called_once_with()`

DavidFair commented on tests/lib/test_openstack_flavor.py:
```diff
@@ -0,0 +1,211 @@
+        return_diff = self.instance.get_missing_flavors(
+            NonCallableMock(), NonCallableMock()
+        )
+        assert [0, 1] == return_diff
```
A small comment here would help visualise why 0, 1. Something alone the lines of 
```
# We expect 0, 1 since the first set contains 0:3 and the second set contains 2:5
# so 0, 1 are the elements missing in set two
```

Feel free to reword this

DavidFair commented on tests/lib/test_openstack_flavor.py:
```diff
@@ -0,0 +1,211 @@
+        mocked_cloud_name = NonCallableMock()
+        flavor_id = NonCallableMock()
+        self.instance.set_flavor_specs(mocked_cloud_name, flavor_id, empty_flavor_spec)
+        self.api.set_flavor_specs.assert_not_called()
```
:+1: I didn't think of this case

DavidFair commented on tests/lib/test_openstack_flavor.py:
```diff
@@ -0,0 +1,211 @@
+        self.api.set_flavor_specs.assert_called_once_with(flavor_id, extra_specs)
+
+    def test_empty_list_migrate_flavors(self):
+        # mock params
+        mocked_source_cloud = NonCallableMock()
+        mocked_dest_cloud = NonCallableMock()
+        # mock return value
+        missing_flavors = NonCallableMock()
+        # mock functions
+        self.instance.get_missing_flavors = Mock()
+        self.instance.get_flavor = Mock()
+        # prepare return value
+        self.instance.get_missing_flavors.return_value = missing_flavors
+        # call method
+        self.instance.get_missing_flavors(mocked_source_cloud, mocked_dest_cloud)
+        # assert at least one function that is called by method
```
We could trim out these comments, since we use the same pattern throughout the code base.
If you'd like to preserve them, maybe a developer section in the README would be a more convenient place to put an example?

meoflynn committed:
Use set comprehension to build sets of flavor names rather than casting lists to set type
Updated comments to refer to source cloud and destination cloud

meoflynn committed:
Update if statement in create_flavor

meoflynn committed:
Updated comments for action tests.
Added assert_called_once_with() to test_list_flavor method to check the api is called correctly for getting a list of flavors
